src/nlp_expr.jl

```diff
@@ -0,0 +1,562 @@
+            for child in reverse(arg.args)
+                push!(stack, (new_ret, child))
+            end
+        elseif arg isa Number
+            push!(parent.args, arg)
+        else
+            push!(parent.args, moi_function(arg))
+        end
+    end
+    return ret
+end
+
+function jump_function(model::Model, f::MOI.ScalarNonlinearFunction)
+    ret = NonlinearExpr(f.head, Any[])
+    stack = Tuple{NonlinearExpr,Any}[]
+    for arg in reverse(f.args)
+        push!(stack, (ret, arg))
+    end
+    while !isempty(stack)
+        parent, arg = pop!(stack)
+        if arg isa MOI.ScalarNonlinearFunction
+            new_ret = NonlinearExpr(arg.head, Any[])
+            push!(parent.args, new_ret)
+            for child in reverse(arg.args)
+                push!(stack, (new_ret, child))
+            end
+        elseif arg isa Number
+            push!(parent.args, arg)
+        else
+            push!(parent.args, jump_function(model, arg))
+        end
+    end
+    return ret
+end
+
+function jump_function_type(::Model, ::Type{<:MOI.ScalarNonlinearFunction})
+    return NonlinearExpr
+end
+
+moi_function_type(::Type{NonlinearExpr}) = MOI.ScalarNonlinearFunction
+
+function constraint_object(c::NonlinearConstraintRef)
+    nlp = nonlinear_model(c.model)
+    data = nlp.constraints[index(c)]
+    return ScalarConstraint(jump_function(c.model, data.expression), data.set)
+end
+
+function jump_function(model::Model, expr::MOI.Nonlinear.Expression)
+    nlp = nonlinear_model(model)
+    parsed = Vector{Any}(undef, length(expr.nodes))
+    adj = MOI.Nonlinear.adjacency_matrix(expr.nodes)
+    rowvals = SparseArrays.rowvals(adj)
+    for i in length(expr.nodes):-1:1
+        node = expr.nodes[i]
+        parsed[i] = if node.type == MOI.Nonlinear.NODE_CALL_UNIVARIATE
+            NonlinearExpr(
+                nlp.operators.univariate_operators[node.index],
+                parsed[rowvals[SparseArrays.nzrange(adj, i)[1]]],
+            )
+        elseif node.type == MOI.Nonlinear.NODE_CALL_MULTIVARIATE
+            NonlinearExpr(
+                nlp.operators.multivariate_operators[node.index],
+                Any[parsed[rowvals[j]] for j in SparseArrays.nzrange(adj, i)],
+            )
+        elseif node.type == MOI.Nonlinear.NODE_MOI_VARIABLE
+            VariableRef(model, MOI.VariableIndex(node.index))
+        elseif node.type == MOI.Nonlinear.NODE_PARAMETER
+            NonlinearParameter(model, node.index)
+        elseif node.type == MOI.Nonlinear.NODE_SUBEXPRESSION
+            NonlinearExpression(model, node.index)
+        elseif node.type == MOI.Nonlinear.NODE_VALUE
+            expr.values[node.index]
+        else
+            # node.type == MOI.Nonlinear.NODE_COMPARISON
+            # node.type == MOI.Nonlinear.NODE_LOGIC
+            error("Unsupported node")
+        end
+    end
+    return parsed[1]
+end
+
+function value(f::Function, expr::NonlinearExpr)
+    return error("TODO")
```

Comment:
I need to think about how to do this. I think it's going to require storing the model in the expression.